src/io/mouse.js

```diff
@@ -1,20 +1,47 @@
 var MathUtil = require('../util/math-util');
 
-function Mouse () {
+function Mouse (runtime) {
     this._x = 0;
     this._y = 0;
     this._isDown = false;
+    /**
+     * Reference to the owning Runtime.
+     * Can be used, for example, to activate hats.
+     * @type{!Runtime}
+     */
+    this.runtime = runtime;
 }
 
 Mouse.prototype.postData = function(data) {
     if (data.x) {
-        this._x = data.x;
+        this._x = data.x - data.canvasWidth / 2;
     }
     if (data.y) {
-        this._y = data.y;
+        this._y = data.y - data.canvasHeight / 2;
     }
     if (typeof data.isDown !== 'undefined') {
         this._isDown = data.isDown;
+        if (this._isDown) {
+            this._activateClickHats(data.x, data.y);
+        }
+    }
+};
+
+Mouse.prototype._activateClickHats = function (x, y) {
```

Comment:
To me it's a question of, how useful do you think the "I/O devices" abstraction is?

Here are some other cases to think about:

- Video sensing: where do you want to do this processing? Specifically something is going to need to receive frames (or some preprocessed data) from outside the worker and map it into some data accessible by block primitives (global video motion for when motion > _, and on-time reporter values).
- Microphone sensing: I think it's the same situation, more or less.
- Variables: we need a place to put variables and lists, which can be global to the project or local to the clone, or cloud variables which require I/O.
- The renderer, the audio engine: as we're discovering with Eric, it would seem good to have a centralized message passing system for these (and presumably cloud variables, etc.)

Regarding your original question: what other things, other than the mouse, are going to produce click hats? I could imagine touch events. If we want to support those, would it be better just to rename Mouse to Pointer or something, and have it handle touch events too? Or just duplicate the click-hat event code in a Touch device? Or should we have one top-level object handling all the logic for creating all these events, as you might suggest?